Docs: Plugin doc review for chunk 1-A (#67070)
grafana/grafana: docs/sources/developers/plugins/_index.md

aliases:
  - ../plugins/developing/
description: Resources for creating Grafana plugins
title: Plugin developer's guide
weight: 200
---

# Grafana plugin developer's guide

> Grafana plugins extend its built-in functionality so that you can accomplish specialized tasks to make Grafana even more useful for them. By making a plugin for your organization, you can connect Grafana to other data sources, ticketing tools, and CI/CD tooling.

Reviewer: ```suggestion
You can extend Grafana's built-in capabilities with plugins. Plugins enable Grafana to accomplish specialized tasks, custom-tailored to your requirements. By making a plugin for your organization, you can connect Grafana to other data sources, ticketing tools, and CI/CD tooling.
```

> You can create plugins for private use or contribute them to the open source community, such as by publishing to the [Grafana plugin catalog](https://grafana.com/grafana/plugins/) alongside hundreds of other community and commercial plugins.

Reviewer: ```suggestion
You can create plugins for private use or contribute them to the open source community by publishing to the [Grafana plugin catalog](https://grafana.com/grafana/plugins/). This catalog has hundreds of other community and commercial plugins.
```

> If you are a Grafana plugin developer or want to become one, this documentation is for you. Use it to get productive quickly, follow along with step-by-step tutorials, or you can refer back to it as you develop.

Reviewer: ```suggestion
If you are a Grafana plugin developer or want to become one, this plugin developer's guide contains the tutorials and reference materials to help you get started. 
```

## Plugin basics

> These are among the types of plugins you can create:

Reviewer: ```suggestion
You can create several types of plugins, including:
```

- **Panel plugins** - Visualize data and navigate between dashboards.
- **Data source plugins** - Link to new databases or other sources of data.
- **App plugins** - Create rich applications for custom out-of-the-box experiences.

> **Note:** To learn more about the types of plugins you can build, refer to the [Plugin management]({{< relref "../../administration/plugin-management" >}}) documentation.

## Contents of this developer's guide

> The following topics give you an idea of what to expect from this guide:

Reviewer: ```suggestion
The following topics are covered in this guide:
```

> - **[Get started with plugins]({{< relref "./get-started-with-plugins" >}})** - The easiest way to start developing Grafana plugins is to use the [Create-plugin](https://www.npmjs.com/package/@grafana/create-plugin) tool.

Reviewer: ```suggestion
- **[Get started with plugins]({{< relref "./get-started-with-plugins" >}})** - Sstart developing Grafana plugins with the [Create-plugin](https://www.npmjs.com/package/@grafana/create-plugin) tool.
```

> - **[Introduction to plugin development]({{< relref "./introduction-to-plugin-development" >}})** - Ground yourself in the fundamentals of Grafana plugin development: frontend and backend development processes, data frames, error handling, and more.

Reviewer: ```suggestion
- **[Introduction to plugin development]({{< relref "./introduction-to-plugin-development" >}})** - Learn the fundamentals of Grafana plugin development: frontend and backend development processes, data frames, error handling, and more.
```

> - **[Create a Grafana plugin]({{< relref "./create-a-grafana-plugin" >}})** - If you're already familiar with plugin creation basics, then it's time to dive deeper into specific development scenarios. You'll find tutorials for creating panel plugins, data source plugins, and more.

Reviewer: ```suggestion
- **[Create a Grafana plugin]({{< relref "./create-a-grafana-plugin" >}})** - If you're familiar with plugin creation, use the tutorials for creating panel plugins, data source plugins, and more to deepen your knowledge.
```

- **[Migrate a plugin]({{< relref "./migration-guide" >}})** - Learn how to upgrade from a previous version of a Grafana plugin, rewrite an old Angular plugin in React, or update to a newer version.
- **[Publish a Grafana plugin]({{< relref "./publish-a-plugin" >}})** - Learn about publishing a plugin to the Grafana plugin catalog, including publishing criteria, packaging, and deployment.
- **[Reference]({{< relref "metadata.md" >}})** - Description of the `plugin.json` schema and plugin metadata.

## Go further

> Learn more about additional tools and find useful examples of a variety of plugin types.

Reviewer: ```suggestion
Learn more about additional tools and see plugin type examples. 
```

aliases:
  - ../plugins/developing/
description: Resources for creating Grafana plugins

> title: Welcome to the Grafana plugin developer's guide

Reviewer 2: Feels overly conversational and verbose for a ToC heading, how about just "Plugin [developer's | development] guide" and then for your h1 below you can have the full "Welcome to the Grafana plugin developer's guide"

Author: Grafana's writer's toolkit defines our tone and voice as "conversational". Nevertheless, I got similar feedback in doc team review and have axed the "Welcome" part.

Reviewer 2: Which is fair for the content, I just expect section titles in ToC to be more matter of fact/task-based so this stuck out a bit for me
<img width="268" alt="Screenshot 2023-04-27 at 11 33 10" src="https://user-images.githubusercontent.com/19860021/234837303-5e82ff45-77b4-4c97-b40c-8cda573d413b.png">

weight: 200
---

# Grafana plugin developer's guide

> Grafana is an open source software project for creating, exploring, and sharing data through dashboards. Plugins extend Grafana's built-in functionality so that users can accomplish specialized tasks to make Grafana even more useful for them. 

Reviewer 2: I probably wouldn't explain Grafana, even briefly, plugin development is a suitably advanced topic that the audience understands this already.

I would maybe include some examples of what plugins do as we have on the grafana intro, specialized tasks feels vague. Example from `/docs/grafana/latest/introduction/`:
> the Grafana OSS plugin framework also enables you to connect other data sources like NoSQL/SQL databases, ticketing tools like Jira or ServiceNow, and CI/CD tooling like GitLab.

Author: Fixed in latest commit.

> Many plugins are available in the [Catalog](https://grafana.com/grafana/plugins/) and others are built by developers for private use or to contribute them to the open source community. 

Reviewer 2: To me this implies that those in the Catalog are not built by developers / contributed to the open source community. 

Suggestion (may need more wordsmithing): Developers can create plugins for private use or contribute them to the open source community, such as by publishing to the Grafana Catalog alongside hundreds of other community and commercial plugins.

Author: Issue is fixed in latest commit.

If you are a Grafana plugin developer or want to become one, this documentation is for you. Use it to get productive quickly, follow along with step-by-step tutorials, or you can refer back to it as you develop.  

## Plugin basics

These are among the types of plugins you can create:

- **Panel plugins** - Visualize data and navigate between dashboards.
- **Data source plugins** - Link to new databases or other sources of data.

> - **App plugins** - Bundle panels and data sources together to create a better user experience.

Reviewer 3: "App Plugins" -> I think the description is not quite correct and I know we are using it somewhere on our website - e.g. https://grafana.com/grafana/plugins/

I would like to emphasise that it's not just bundling panels and datasources together but you can do way more. It is basically a custom react + backend application running inside Grafana and you are free to do whatever you need in there. Some examples of App Plugins Grafana Labs has built:
- https://play.grafana.org/a/k6-app
- https://play.grafana.org/a/grafana-incident-app/incidents
- https://play.grafana.org/a/grafana-ml-app/home
- https://play.grafana.org/a/grafana-k8s-app/cluster

 I would say something like

"**App plugins** - Create rich applications for custom out-of-the-box experience"

Reviewer 2: Agree with the above

> **Note:** To learn more about the types of plugins you can build, refer to the [Plugin Overview]({{< relref "../../administration/plugin-management/" >}}) documentation.

## Contents of this developer's guide

The following topics give you an idea of what to expect from this guide:

> - **[Get started with plugins](get-started-with-plugins/)** - The easiest way to start developing Grafana plugins is to use the [Create-plugin](https://www.npmjs.com/package/@grafana/create-plugin) tool. 

Reviewer 2: maybe worth merging get started and create a plugin?

Author: That's worth considering. What I have in mind with all the landing pages (like "get started") is that they will be principally a TOC for the section. I may reconsider when I get to that step in the writing process though.

> - **[Introduction to plugin development](introduction-to-plugin-development/)** - Ground yourself by learning the key concepts involved in Grafana plugin development: frontend and backend development processes, data frames, error handling, and more. 

Reviewer 2: I see the use of Fundamentals elsewhere in the docs, maybe for consistency we should use that?

Author: Fixed in latest commit.

- **[Create a Grafana plugin](create-a-grafana-plugin)** - If you're already familiar with plugin creation basics, then it's time to dive deeper into specific development scenarios. You'll find tutorials for creating panel plugins, data source plugins, and more. 
- **[Migrate a plugin](migration-guide)** - Learn how to upgrade from a previous version of a Grafana plugin, rewrite an old Angular plugin in React, or update to a newer version. 

> - **[Publish a Grafana plugin](publish-a-plugin/)**  -  Learn about publishing a plugin to the Grafana plugin catalog, including publishing criteria, signing process, and publishing.

Author: Issue is fixed in latest commit.

- **[Legacy plugins](legacy/)** - If you find yourself needing to work with Legacy plugins, check here for recommendations.
- **[Reference](reference/)** - Description of the `plugin.json` schema and plugin metadata.

## Go further

Learn more about additional tools and find useful examples of a variety of plugin types.

### User interface creation

Explore the many UI components in our [Grafana UI library](https://developers.grafana.com/ui).

### Code examples

> For inspiration, check out our [plugin examples](https://github.com/grafana/grafana-plugin-examples).

Author: Issue is fixed in latest commit.

weight: 200
---

> # Welcome to the Grafana plugin developer's guide

Reviewer: I scanned through our other docs and didn't see another page title with "Welcome to the..." in it. Maybe there is another title that would work like "Grafana plugin developer's guide"?

Author: Oh gee. Honestly, my preference would be for us to add "Welcome to the... to all the other doc pages. Better for Grafana's personality, I think.

Until that happens, I think you speak wisely.

> Grafana is an open source software project for creating, exploring, and sharing all of your data through beautiful, flexible dashboards. Plugins extend its built-in functionality so that users can accomplish specialized tasks to make Grafana even more useful for them. 

Reviewer: ```suggestion
Grafana is an open source software project for creating, exploring, and sharing data through dashboards. Plugins extend Grafana's built-in functionality so that users can accomplish specialized tasks. 
```

Author: nice suggestions, I will check in fix

Many plugins are available in the [Catalog](https://grafana.com/grafana/plugins/) and others are built by developers for private use or to contribute them to the open source community. 

If you are a Grafana plugin developer or want to become one, this documentation is for you. Use it to get productive quickly, follow along with step-by-step tutorials, or you can refer back to it as you develop.  

## Plugin basics

These are among the types of plugins you can create:

> - **Panel plugins** - Visualize data and navigating between dashboards.

Reviewer: ```suggestion
- **Panel plugins** - Visualize data and navigate between dashboards.
```

Author: Will check in fix soon

- **Data source plugins** - Link to new databases or other sources of data.
- **App plugins** - Bundle panels and data sources together to create a better user experience.

> **Note:** To learn more about the types of plugins you can build, refer to the [Plugin Overview]({{< relref "../../administration/plugin-management/" >}}) documentation.

## Contents of this developer's guide

The following topics give you an idea of what to expect from this guide:

> - **[Get started with plugins](get-started-with-plugins/)** - The easiest way to start developing Grafana plugins is to use the [Create-plugin](https://www.npmjs.com/package/@grafana/create-plugin) tool. Use it to get your plugin up and running as fast as possible. 

Reviewer: Removed the last sentence since it's duplicating content from the first sentence. 
```suggestion
- **[Get started with plugins](get-started-with-plugins/)** - The easiest way to start developing Grafana plugins is to use the [Create-plugin](https://www.npmjs.com/package/@grafana/create-plugin) tool.  
```

Author: will check in fix

> - **[Introduction to plugin development](introduction-to-plugin-development/)** - Ground yourself by learning the key concepts involved in Grafana plugin development. For example, learn about frontend and backend development processes, data frames, and error handling. 

Reviewer: ```suggestion
- **[Introduction to plugin development](introduction-to-plugin-development/)** - Learn the key concepts involved in Grafana plugin development: frontend and backend development processes, data frames, and error handling. 
```

Author: Will check in fix

- **[Create a Grafana plugin](create-a-grafana-plugin)** - If you're already familiar with plugin creation basics, then it's time to dive deeper into specific development scenarios. You'll find tutorials for creating panel plugins, data source plugins, and more. 
- **[Migrate a plugin](migration-guide)** - Learn how to upgrade from a previous version of a Grafana plugin, rewrite an old Angular plugin in React, or update to a newer version. 

> - **[Publish a Grafana plugin](publish-a-plugin/)**  - Here you'll find what you need to know if you want to publish a plugin to the Grafana Catalog, including publishing criteria, signing process, and publishing.

Reviewer: ```suggestion
- **[Publish a Grafana plugin](publish-a-plugin/)** - Learn about publishing a plugin to the Grafana Catalog, including publishing criteria, signing process, and publishing.
```

weight: 200
---

# Welcome to the Grafana plugin developer's guide

Grafana is an open source software project for creating, exploring, and sharing all of your data through beautiful, flexible dashboards. Plugins extend its built-in functionality so that users can accomplish specialized tasks to make Grafana even more useful for them. 

Many plugins are available in the [Catalog](https://grafana.com/grafana/plugins/) and others are built by developers for private use or to contribute them to the open source community. 

If you are a Grafana plugin developer or want to become one, this documentation is for you. Use it to get productive quickly, follow along with step-by-step tutorials, or you can refer back to it as you develop.  

## Plugin basics

These are among the types of plugins you can create:

- **Panel plugins** - Visualize data and navigating between dashboards.
- **Data source plugins** - Link to new databases or other sources of data.
- **App plugins** - Bundle panels and data sources together to create a better user experience.

> - **Other plugins** - Visualize data source logs or streaming data sources.

Author: deleted Other bullet point

weight: 200
---

# Welcome to the Grafana plugin developer's guide

Grafana is an open source software project for creating, exploring, and sharing all of your data through beautiful, flexible dashboards. Plugins extend its built-in functionality so that users can accomplish specialized tasks to make Grafana even more useful for them. 

Many plugins are available in the [Catalog](https://grafana.com/grafana/plugins/) and others are built by developers for private use or to contribute them to the open source community. 

If you are a Grafana plugin developer or want to become one, this documentation is for you. Use it to get productive quickly, follow along with step-by-step tutorials, or you can refer back to it as you develop.  

## Plugin basics

These are among the types of plugins you can create:

- **Panel plugins** - Visualize data and navigating between dashboards.
- **Data source plugins** - Link to new databases or other sources of data.
- **App plugins** - Bundle panels and data sources together to create a better user experience.
- **Other plugins** - Visualize data source logs or streaming data sources.

> The possibilities are as endless as your imagination, and you can use this guide to help make your dreams become reality. 

Author: On second thought, delete this sentence as it is inconsistent with our voice and tone.

   Note: To learn more about the types of plugins you can build, refer to the [Plugin Overview]({{< relref "../../administration/plugin-management/" >}}) documentation.

## Contents of this developer's guide

The following topics give you an idea of what to expect from this guide:

> - [**Get started with plugins** - The easiest way to start developing Grafana plugins is to use the Create-plugin tool. Walk step by step for downloading the tool and using it to get your plugin up and running as fast as possible. 

Reviewer 4: Apart from the missing link, I would format `create-plugin` (lower-case and with code format) since it's a CLI tool. Also, seems like the second sentence is missing something?

Author: @andresmgot - The missing link will be corrected with my next check-in. Regarding the name of the tool, we discussed this topic in the doc team and decided that to be consistent across all Grafana products we would be allowing tool names to be identical with CLI commands, but the tool name would use an initial cap. This makes it easier to tell whether the text is talking about a tool or a command. Regarding the end of the sentence, yikes, I shortened it. The next check-in should be better.

Author: Checked in fix, available for review.

Reviewer 4: Makes sense. In this case, I believe that the tool name is "[Create Plugin](https://github.com/grafana/plugin-tools/tree/main/packages/create-plugin)" while the command version of it would be `create-plugin`.

Author: Personally I like "Create Plugin" as a name, however the doc team's style is to not use Title Case. For the names of CLI tools, we are using the command name exactly with a cap at the start.

> - [**Introduction to plugin development** - Ground yourself by learning the key concepts involved in Grafana plugin development. For example, learn about frontend and backend development processe, data frames, and error handling. 

Author: Thanks. Will be fixed soon.

> - [**Introduction to plugin development** - Ground yourself by learning the key concepts involved in Grafana plugin development. For example, learn about frontend and backend development processe, data frames, and error handling. 

Author: Checked in fix, available for review.

- [**Create a Grafana plugin** - If you are already familiar with plugin creation basics, then it's time to dive deeper into specific development scenarios. You will find tutorials for creating panel plugins, data source plugins, and more. 
- [**Migrate a plugin** - Learn how to upgrade from a previous version of a Grafana plugin, rewrite an old Angular plugin in React, or update to a newer version. 
- [**Publish a Grafana plugin**  - Here you'll find what you need to know if you want to publish a plugin to the Grafana Catalog, including publishing criteria, signing process, and publishing.
- [**Advanced topics** - Learn all about advanced plugin features such as nested plugins. If you find yourself needing to work with Legacy plugins, check here for recommendations.

> - [**Reference** - Description of the Plugin.json schema and plugin metadata.

Reviewer: Are the bolded words in these bullets supposed to be links? The previous deleted bullets had links. Are the opening [ left-overs?

Author: I will soon check in an update that will make all of these links. I had to build the files and add links to those files.
